tools/fuzzer/main.c

```diff
@@ -0,0 +1,337 @@
+	case SOF_IPC_GLB_STREAM_MSG:
+	case SOF_IPC_GLB_TRACE_MSG:
+		fuzzer->platform->mailbox_read(fuzzer, 0, &r, sizeof(r));
+		break;
+	default:
+		fprintf(stderr, "error: unknown DSP message 0x%x\n", cmd);
+		break;
+	}
+}
+
+/* called by platform when it receives IPC message reply */
+void fuzzer_ipc_msg_reply(struct fuzz *fuzzer)
+{
+	int ret;
+
+	ret = fuzzer->platform->get_reply(fuzzer, &fuzzer->msg);
+	if (ret < 0)
+		fprintf(stderr, "error: incorrect DSP reply\n");
+
+	ipc_dump(&fuzzer->msg);
+
+	pthread_mutex_lock(&ipc_mutex);
+	pthread_cond_signal(&ipc_cond);
+	pthread_mutex_unlock(&ipc_mutex);
+}
+
+/* called by platform when FW crashses */
+void fuzzer_ipc_crash(struct fuzz *fuzzer, unsigned int offset)
+{
+	/* TODO: DSP FW has crashed. dump stack, regs, last IPC, log etc */
+	fprintf(stderr, "DSP FW crash\n");
+	exit(EXIT_FAILURE);
+}
+
+int fuzzer_send_msg(struct fuzz *fuzzer)
+{
+	struct timespec timeout;
+	struct timeval tp;
+	int ret;
+
+	ipc_dump(&fuzzer->msg);
+
+	/* send msg */
+	ret = fuzzer->platform->send_msg(fuzzer, &fuzzer->msg);
+	if (ret < 0) {
+		fprintf(stderr, "error: message tx failed\n");
+		return ret;
+	}
+
+	/* wait for ipc reply */
+	gettimeofday(&tp, NULL);
+	timeout.tv_sec  = tp.tv_sec;
+	timeout.tv_nsec = tp.tv_usec * 1000;
+	timeout.tv_nsec += 300000000; /* 300ms timeout */
+
+	/* first lock the boot wait mutex */
+	pthread_mutex_lock(&ipc_mutex);
+
+	/* now wait for mutex to be unlocked by boot ready message */
+	ret = pthread_cond_timedwait(&ipc_cond, &ipc_mutex, &timeout);
+	if (ret == ETIMEDOUT) {
+		ret = -EINVAL;
+		fprintf(stderr, "error: IPC timeout\n");
+		ipc_dump_err(&fuzzer->msg);
+		pthread_mutex_unlock(&ipc_mutex);
+		exit(0);
+	}
+
+	pthread_mutex_unlock(&ipc_mutex);
+
+	/*
+	 * sleep for 5 ms before continuing sending the next message.
+	 * This helps with the condition signaling. Without this,
+	 * the condition seems to always satisfy and
+	 * the fuzzer never waits for a response from the DSP.
+	 */
+	usleep(50000);
+
+	return ret;
+}
+
+int main(int argc, char *argv[])
+{
+	struct fuzz fuzzer;
+	int ret;
+	char opt;
+	char *topology_file;
+	char *platform_name = NULL;
+	int i, j;
+	int regions = 0;
+
+	/* parse arguments */
+	while ((opt = getopt(argc, argv, "ht:p:")) != -1) {
+		switch (opt) {
+		case 't':
+			topology_file = optarg;
+			break;
+		case 'p':
+			platform_name = optarg;
+			break;
+		case 'h':
+			usage(argv[0]);
+			exit(0);
+			break;
+		default:
+			usage(argv[0]);
+			exit(EXIT_FAILURE);
+		}
+	}
+
+	/* initialise emulated target device */
+	if (!platform_name) {
+		fprintf(stderr, "error: no target platform specified\n");
+		usage(argv[0]);
+		exit(EXIT_FAILURE);
+	}
+
+	/* find platform */
+	for (i = 0; i < ARRAY_SIZE(platform); i++) {
+		if (!strcmp(platform[i]->name, platform_name))
+			goto found;
+	}
+
+	/* no platform found */
+	fprintf(stderr, "error: platform %s not supported\n", platform_name);
+	usage(argv[0]);
+	exit(EXIT_FAILURE);
+
+found:
```

Comment:
This is just a test tool, but this goto-usage seems a bit unnecessary. You could use a boolean or split main() into a few subfunctions. Not a showstopper, upto to you...